service/view_building_coordinator.cc

```diff
@@ -0,0 +1,253 @@
+
+public:
+    view_building_coordinator(abort_source& as, replica::database& db, raft_group0& group0, db::system_keyspace& sys_ks, const topology_state_machine& topo_sm) 
+        : _db(db)
+        , _group0(group0)
+        , _sys_ks(sys_ks)
+        , _topo_sm(topo_sm)
+        , _as(as) 
+        , _vb_state(std::nullopt)
+    {}
+
+    future<> run();
+
+    virtual void on_create_view(const sstring& ks_name, const sstring& view_name) override { _cond.broadcast(); }
+    virtual void on_update_view(const sstring& ks_name, const sstring& view_name, bool columns_changed) override {}
+    virtual void on_drop_view(const sstring& ks_name, const sstring& view_name) override { _cond.broadcast(); }
+
+private:
+    future<group0_guard> start_operation() {
+        auto guard = co_await _group0.client().start_operation(_as);
+        co_return std::move(guard);
+    }
+
+    future<> await_event() {
+        _as.check();
+        co_await _cond.when();
+        vbc_logger.debug("event awaited");
+    }
+
+    future<> initialize_coordinator_state() {
+        auto tasks = co_await _sys_ks.get_view_building_coordinator_tasks();
+        _vb_state = view_building_state {
+            .build_tasks = std::move(tasks)
+        };
+    }
+
+    future<> update_coordinator_state(group0_guard guard);
+
+    future<> add_view(const view_name& view_name, view_building_state& state_copy, group0_batch& batch);
+    future<> remove_view(const view_name& view_name, view_building_state& state_copy, group0_batch& batch);
+
+private:
+    future<std::set<view_name>> load_all_views() {
+        static const sstring query = format("SELECT keyspace_name, view_name FROM {}.{}", db::schema_tables::v3::NAME, db::schema_tables::v3::VIEWS);
+
+        std::set<view_name> views;
+        co_await _sys_ks.query_processor().query_internal(query, [&] (const cql3::untyped_result_set_row& row) -> future<stop_iteration> {
+            auto ks_name = row.get_as<sstring>("keyspace_name");
+            auto view_name = row.get_as<sstring>("view_name");
+
+            views.insert({ks_name, view_name});
+            co_return stop_iteration::no;
+        });
+        co_return views;
+    }
+
+    future<std::set<view_name>> load_built_views() {
+        auto built_views = co_await _sys_ks.load_built_views();
+        co_return std::set<view_name>(std::make_move_iterator(built_views.begin()), std::make_move_iterator(built_views.end()));
+    }
+};
+
+future<> view_building_coordinator::run() {
+    auto abort = _as.subscribe([this] noexcept {
+        _cond.broadcast();
+    });
+
+    co_await initialize_coordinator_state();
+
+    while (!_as.abort_requested()) {
+        vbc_logger.debug("coordinator loop iteration");
+        try {
+            co_await update_coordinator_state(co_await start_operation());
+
+            // TODO
+            // Do actual work, send RPCs to build a particular view's range
+            co_await await_event();
```

Comment:
I might be missing something (and I see the logic here is still only in draft stage) but shouldn't we wait for an event before starting a group0 operation?